In [1]:
from pprint import pprint

fname = r'input.txt'
# fname = r'test.txt'

# Load input
with open(fname, 'r') as fp:
    lines = fp.readlines()

garden_coords = set()
for ln_num, ln in enumerate(lines):
    for char_num, char in enumerate(ln.strip()):
        if char == '.':
            garden_coords.add((ln_num, char_num))
        
        if char == 'S':
            start_pos = (ln_num, char_num)
max_x = ln_num + 1
max_y = char_num + 1

In [2]:
garden_coords.add(start_pos)
visited_gardens = {start_pos}

In [3]:
processed_gardens = set()
possible_end_gardens = set()
# num_steps = 26501365


In [4]:
def explore_garden(num_steps, start_pos):
    visited_gardens = {start_pos}
    for _ in range(num_steps):  # take 64 steps
        gardens_this_step = set()
        for x, y in visited_gardens:  # go through each garden visited
            for new_coords in [(x+1, y), (x-1, y), (x, y+1), (x, y-1)]: # try one step in any direction
                    if new_coords in garden_coords:  # is it a garden?
                        gardens_this_step.add(new_coords)  # mark it as visited
        visited_gardens=gardens_this_step

    return len(visited_gardens)


In [5]:
def number_gardens_with_repeats(num_steps, start_pos):
    visited_gardens = {start_pos}
    processed_gardens = set()
    possible_end_gardens = set()
    for step_num in range(num_steps):
        gardens_this_step = set()
        for x, y in visited_gardens:  # go through each garden visited
            for new_coords in [(x+1, y), (x-1, y), (x, y+1), (x, y-1)]: # try one step in any direction
                    new_x, new_y = new_coords
                    if (new_x % max_x, new_y % max_y) in garden_coords:  # is it a garden?
                        if new_coords not in processed_gardens:
                            gardens_this_step.add(new_coords)  # mark it as visited
        processed_gardens.update(gardens_this_step)
        if step_num % 2 != num_steps % 2:
            possible_end_gardens.update(gardens_this_step)
        visited_gardens = gardens_this_step
    
    return(len(possible_end_gardens))

# Check for test gardens

In [6]:
# Try on test gardens

# for num_steps in [6, 10, 50, 100, 500, 1000, 5000]:
#     print(f'{num_steps=}  num_gardens={number_gardens_with_repeats(num_steps, start_pos)}')


## Results
When run on test gardens, results match given values on test input:
```
num_steps=6  num_gardens=16
num_steps=10  num_gardens=50
num_steps=50  num_gardens=1594
num_steps=100  num_gardens=6536
num_steps=500  num_gardens=167004
num_steps=1000  num_gardens=668697
num_steps=5000  num_gardens=16733044
```

In [9]:
import numpy as np

In [38]:
65 + 131*202300

26501365

In [44]:
num_visted_gardens = []
for i in range(10):
    num_steps = 65 + 131*i
    num_visted_gardens.append(number_gardens_with_repeats(num_steps, start_pos))

num_visted_gardens

[3882, 34441, 95442, 186885, 308770, 461097, 643866, 857077, 1100730, 1374825]

In [45]:
arr = np.array(num_visted_gardens)
arr

array([   3882,   34441,   95442,  186885,  308770,  461097,  643866,
        857077, 1100730, 1374825])

In [58]:
print(arr)
print(np.diff(arr))
print(np.diff(np.diff(arr)))
print(np.diff(np.diff(np.diff(arr))))

[   3882   34441   95442  186885  308770  461097  643866  857077 1100730
 1374825]
[ 30559  61001  91443 121885 152327 182769 213211 243653 274095]
[30442 30442 30442 30442 30442 30442 30442 30442]
[0 0 0 0 0 0 0]


## Figure out quadratic equation

$$
\begin{darray}{rcl}
f(x) &=& ax^2 + bx + c \\
f(0) &=& c \\
c &=& 3882 \\
\\
f'(x) &=& 2ax + b \\
f''(x) &=& 2a \\
f''(\cdot) &=& 30442 \\
a &=& 15221 \\
\\
f(1) &=& 34441 \\
15221(1)^2 + b + 3882 &=& 34441 \\
b &=& 15338 \\
\\
f(x) &=& 15221 x^2 + 15338 x + 3882\\
\end{darray}
$$

## Testing quadratic equation

In [60]:
def num_gardens_visited(num_reps):
    return 15221 * num_reps**2 + 15338 * num_reps + 3882

In [63]:
# Test
[(arr[i], num_gardens_visited(i)) for i in range(10)]

[(3882, 3882),
 (34441, 34441),
 (95442, 95442),
 (186885, 186885),
 (308770, 308770),
 (461097, 461097),
 (643866, 643866),
 (857077, 857077),
 (1100730, 1100730),
 (1374825, 1374825)]

In [64]:
num_gardens_visited(202300)

622926941971282

# Ignore information below
Stuff that didn't really work but I'm not ready to delete yet.

# Analysis

By inspection, can reach all 7645 accessible gardens in one map repeat with 130 steps (accounting for parity).

Investigating starting at edge:

In [ ]:
for num_steps in [256, 258, 260, 262]:
    num_gardens = explore_garden(num_steps, (0,0))
    print(f'{num_steps=}, {num_gardens=}')

Investing odd parity

In [ ]:
for num_steps in [255, 257, 259, 261]:
    num_gardens = explore_garden(num_steps, (0,0))
    print(f'{num_steps=}, {num_gardens=}')

# Analysis

Can reach all 7576 odd-partiy gardens in 259 steps

In [ ]:
number_steps = 26501365

# number of steps once at edge of first garden = 65
print(remaining_steps:=number_steps - 65)
print(possible_map_explorations:=remaining_steps / 131)
